### PR TITLE
Java Loggers

### PR DESCRIPTION
**Pull request by TheMellyBee**
## Description
This incremental iteration focuses on the Java loggers, but note that the TransientLogger is not done until I do the Writer (next iteration).  Please remember while reviewing that the metric we have right now is the Integral Metric going from start to end. 

NOTE: This was built off of a branch and then another branch due to PRs. Sadly that's made the commits pretty wild, only the last 10 or so are from this feature and all will be squashed on merge. I did research how to fix that but it looks like best way I saw was on the squash :/

- [ X] I have reviewed the [Guidelines for Contributing](CONTRIBUTING.md) and the [Code of Conduct](CODE_OF_CONDUCT.md).

### CODE REVIEW

File: java/core/src/main/java/com/whylogs/api/logger/Logger.java [hunk 1]
```diff
@@ -1,3 +1,85 @@
...
+  protected abstract ArrayList<DatasetProfile> getMatchingProfiles(Object data);
+
+  protected abstract <O> ArrayList<DatasetProfile> getMatchingProfiles(Map<String, O> data);
+
+  @Override
+  public void close() {
+    isClosed = true;
+  }
+
+  public ResultSet log(HashMap<String, Object> data) {
```
Comment (richard-rogers) [inline java/core/src/main/java/com/whylogs/api/logger/Logger.java:66]: I assume this corresponds to the Python `log(row=...)` and the other data formats will come later?

File: java/core/src/main/java/com/whylogs/api/logger/rollingLogger/TimedRollingLogger.java [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+  private String fileExtension;
+  private int interval;
+  private Character when = 'H'; // TODO: Make the Literals of S M H D
+  private boolean utc = false;
+  private boolean align = true;
+  private boolean skipEmpty = false;
+  private String suffix;
+
+  private DatasetProfile currentProfile;
+  private Callable<Writer> callback; // TODO: this isn't the write signature
```
Comment (richard-rogers) [inline java/core/src/main/java/com/whylogs/api/logger/rollingLogger/TimedRollingLogger.java:25]: write or right? :)

File: java/core/src/main/java/com/whylogs/api/logger/rollingLogger/TimedRollingLogger.java [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+      case 'H':
+        this.interval = 60 * 60; // one hour
+        this.suffix = "%Y-%m-%d_%H";
+        break;
+      case 'D':
+        this.interval = 60 * 60 * 24; // one day
+        this.suffix = "%Y-%m-%d";
+        break;
+      default:
+        throw new IllegalArgumentException("Invalid value for when: " + this.when);
```
Comment (richard-rogers) [inline java/core/src/main/java/com/whylogs/api/logger/rollingLogger/TimedRollingLogger.java]: Perhaps include the valid values?

File: java/core/src/main/java/com/whylogs/api/logger/rollingLogger/TimedRollingLogger.java [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+    }
+
+    this.scheduler = new Scheduler(initialRunAfter, this.interval, this::doRollover, null);
+    this.scheduler.start();
+
+    // autocloseable closes at end
+  }
+
+  private int computeCurrentBatchTimestamp(long nowEpoch) {
+    int roundedNow = (int) nowEpoch;
```
Comment (richard-rogers) [inline java/core/src/main/java/com/whylogs/api/logger/rollingLogger/TimedRollingLogger.java]: Comment what line 108 is doing?

File: java/core/src/main/java/com/whylogs/api/writer/Writable.java [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+
+  static FileWriter safeOpenWrite(String path) {
+    // Open 'path' for writing, creating any parent directories as needed
+    File file = new File(path);
+    FileWriter writer = null;
+    try {
+      writer = new FileWriter(file, true);
+    } catch (Exception e) {
+      System.out.println("Error: " + e);
+      e.printStackTrace();
```
Comment (richard-rogers) [inline java/core/src/main/java/com/whylogs/api/writer/Writable.java:16]: Do we want to do this?

File: java/core/src/main/java/com/whylogs/api/writer/Writer.java [hunk 6]
```diff
@@ -0,0 +1,17 @@
+package com.whylogs.api.writer;
+
+import java.util.Optional;
+
+// TODO: this is a temp holding class for logger that will be implmented next
+public abstract class Writer {
+  /*Validate an interval configuration for a given writer.
+  Some writer only accepts certain interval configuration. By default, this should return True for a valid
+  non-negative interval.*/
+  public void check_interval(int interval_seconds) {
```
Comment (richard-rogers) [inline java/core/src/main/java/com/whylogs/api/writer/Writer.java:10]: Why is this in the base class? (same question for Python)

**Review by richard-rogers — DISMISSED** [review]

File: java/core/src/main/java/com/whylogs/api/logger/rollingLogger/TimedRollingLogger.java (hunk 4 above)
Comment (TheMellyBee) [reply java/core/src/main/java/com/whylogs/api/logger/rollingLogger/TimedRollingLogger.java]: Sure :)

File: java/core/src/main/java/com/whylogs/api/logger/rollingLogger/TimedRollingLogger.java (hunk 2 above)
Comment (TheMellyBee) [reply java/core/src/main/java/com/whylogs/api/logger/rollingLogger/TimedRollingLogger.java:25]: hahahah both... definitely both

File: java/core/src/main/java/com/whylogs/api/logger/Logger.java (hunk 1 above)
Comment (TheMellyBee) [reply java/core/src/main/java/com/whylogs/api/logger/Logger.java:66]: Correct :)

File: java/core/src/main/java/com/whylogs/api/logger/rollingLogger/TimedRollingLogger.java (hunk 3 above)
Comment (TheMellyBee) [reply java/core/src/main/java/com/whylogs/api/logger/rollingLogger/TimedRollingLogger.java]: Sounds great!

File: java/core/src/main/java/com/whylogs/api/writer/Writable.java (hunk 5 above)
Comment (TheMellyBee) [reply java/core/src/main/java/com/whylogs/api/writer/Writable.java:16]: No this is just temporary while getting a basic holder up.

File: java/core/src/main/java/com/whylogs/api/writer/Writer.java (hunk 6 above)
Comment (TheMellyBee) [reply java/core/src/main/java/com/whylogs/api/writer/Writer.java:10]: That's fair. I'm curious to that as well. I will explore that further in the next iteration. Right now it was just a holder

**Review by richard-rogers — APPROVED** [review]

**Pull request merged**